core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java

```diff
@@ -69,12 +75,31 @@ public class RESTCatalog implements Catalog, SupportsNamespaces, Configurable<Co
   private Object conf = null;
   private FileIO io = null;
 
+  public RESTCatalog() {
+    this(new HTTPClientFactory());
+  }
+
   RESTCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
     this.clientBuilder = clientBuilder;
   }
 
   @Override
   public void initialize(String name, Map<String, String> props) {
+    RESTCatalogConfigResponse config = fetchConfig(props);
+    Map<String, String> mergedProps = config.merge(props);
+    initializePostConfig(name, mergedProps);
+  }
+
+  /**
+   * Performs the normal initialization steps after calling the config route and merging the initial
+   * configuration with any configuration provided by the server.
+   * <p>
+   * See {@link RESTCatalogConfigResponse} for details on merging the initial client and server configuration.
+   *
+   * @param name  Name of this RESTCatalog
+   * @param props Properties to use, merged with any server-provided configuration.
+   */
+  private void initializePostConfig(String name, Map<String, String> props) {
```

Comment:
My main reasoning for making it a separate function was so that the non-merged properties didn't get reused accidentally. I can merge it into one function if we prefer.